internal/internal_activity.go

```diff
@@ -302,98 +302,131 @@ func isActivityContext(inType reflect.Type) bool {
 	return inType != nil && inType.Implements(contextElem)
 }
 
-func validateFunctionAndGetResults(f interface{}, values []reflect.Value, dataConverter converter.DataConverter) (*commonpb.Payloads, error) {
-	resultSize := len(values)
-
-	if resultSize < 1 || resultSize > 2 {
-		fnName, _ := getFunctionName(f)
-		return nil, fmt.Errorf(
-			"the function: %v signature returns %d results, it is expecting to return either error or (result, error)",
-			fnName, resultSize)
+func setActivityParametersIfNotExist(ctx Context) Context {
+	params := getActivityOptions(ctx)
+	var newParams ExecuteActivityOptions
+	if params != nil {
+		newParams = *params
+		if params.RetryPolicy != nil {
+			newRetryPolicy := *newParams.RetryPolicy
+			newParams.RetryPolicy = &newRetryPolicy
+		}
 	}
+	return WithValue(ctx, activityOptionsContextKey, &newParams)
+}
 
-	var result *commonpb.Payloads
+func setLocalActivityParametersIfNotExist(ctx Context) Context {
+	params := getLocalActivityOptions(ctx)
+	var newParams ExecuteLocalActivityOptions
+	if params != nil {
+		newParams = *params
+	}
+	return WithValue(ctx, localActivityOptionsContextKey, &newParams)
+}
 
-	// Parse result
-	if resultSize > 1 {
-		retValue := values[0]
+type activityEnvironmentInterceptor struct {
+	env                 *activityEnvironment
+	inboundInterceptor  ActivityInboundInterceptor
+	outboundInterceptor ActivityOutboundInterceptor
+	fn                  interface{}
+}
 
-		var ok bool
-		if result, ok = retValue.Interface().(*commonpb.Payloads); !ok {
-			if retValue.Kind() != reflect.Ptr || !retValue.IsNil() {
-				var err error
-				if result, err = encodeArg(dataConverter, retValue.Interface()); err != nil {
-					return nil, err
-				}
-			}
-		}
+func getActivityEnvironmentInterceptor(ctx context.Context) *activityEnvironmentInterceptor {
+	a := ctx.Value(activityEnvInterceptorContextKey)
+	if a == nil {
+		panic("getActivityEnvironmentInterceptor: Not an activity context")
 	}
+	return a.(*activityEnvironmentInterceptor)
```

Comment:
need to check type here?

Reply:
No more than we do in any other context `Value` lookup where we control the key and the setting internally and therefore know the type